src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -111,74 +154,164 @@ internal abstract class AbstractAddParameterCodeFixProvider<
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFacts,
+            ImmutableArray<IMethodSymbol> methodCandidates,
+            SeparatedSyntaxList<TArgumentSyntax> arguments,
+            TArgumentSyntax argumentOpt,
+            SeparatedSyntaxList<TTypeSyntax> typeArguments)
+        {
+            var methodsAndArgumentToAdd = ArrayBuilder<(IMethodSymbol constructor, TArgumentSyntax argument, int index, TTypeSyntax typeArgument)>.GetInstance();
 
-            foreach (var constructor in type.InstanceConstructors.OrderBy(m => m.Parameters.Length))
+            foreach (var method in methodCandidates.OrderBy(m => m.Parameters.Length))
             {
-                if (constructor.IsNonImplicitAndFromSource() &&
-                    NonParamsParameterCount(constructor) < arguments.Count)
+                if (method.IsNonImplicitAndFromSource())
                 {
-                    var argumentToAdd = DetermineFirstArgumentToAdd(
-                        semanticModel, syntaxFacts, comparer, constructor,
-                        arguments, argumentOpt);
+                    var argumentToAdd = GetArgumentInsertPositionForMethodCandidate(semanticModel, syntaxFacts, method, arguments, argumentOpt);
+                    var typeArgumentToAdd = GetTypeArgumentForMethodCandidate(method, typeArguments);
 
-                    if (argumentToAdd != null)
+                    if (argumentToAdd != null || typeArgumentToAdd != null)
                     {
-                        if (argumentOpt != null && argumentToAdd != argumentOpt)
-                        {
-                            // We were trying to fix a specific argument, but the argument we want
-                            // to fix is something different.  That means there was an error earlier
-                            // than this argument.  Which means we're looking at a non-viable 
-                            // constructor.  Skip this one.
-                            continue;
-                        }
-
-                        constructorsAndArgumentToAdd.Add(
-                            (constructor, argumentToAdd, arguments.IndexOf(argumentToAdd)));
+                        methodsAndArgumentToAdd.Add(
+                            (method, argumentToAdd, arguments.IndexOf(argumentToAdd), typeArgumentToAdd));
                     }
                 }
             }
 
-            // Order by the furthest argument index to the nearest argument index.  The ones with
-            // larger argument indexes mean that we matched more earlier arguments (and thus are
-            // likely to be the correct match).
-            foreach (var tuple in constructorsAndArgumentToAdd.OrderByDescending(t => t.index))
+            return methodsAndArgumentToAdd.ToImmutableArray();
+        }
+
+        private TTypeSyntax GetTypeArgumentForMethodCandidate(IMethodSymbol method, SeparatedSyntaxList<TTypeSyntax> typeArguments)
```

Comment:
Should this be called "GetPossibleTypeArgumentToAddToMethod"?

Naming mismatch.  "ForMethodCandidate" but "IMethodSymbol method".  Please try to keep naming synced between all these concepts.